Do not trigger stream configuration updates in case user is capturing screen in large presenter overlay mode

youennf opened the pull request:
#### 88ff63ad3bde35abf2ae32a3136cbe2ac6289dc0
<pre>
Do not trigger stream configuration updates in case user is capturing screen in large presenter overlay mode

Do not trigger stream configuration updates in case user is capturing screen in large presenter overlay mode
<a href="https://bugs.webkit.org/show_bug.cgi?id=273684">https://bugs.webkit.org/show_bug.cgi?id=273684</a>
<a href="https://rdar.apple.com/125925090">rdar://125925090</a>

Reviewed by Eric Carlson.

We enter in a loop of reconfiguration when trying to update the stream configuration size when user selects large presenter overlay mode.
We are now skipping the reconfiguration step.
This reintroduces black stripes like there used to have before <a href="https://rdar.apple.com/124131045">rdar://124131045</a>.
A future patch will fix this by adding cropping within ScreenCaptureKitCaptureSource.

We detect large presenter overlay mode by:
- detecting whether video effect is enabled (which means presenter mode is on or off)
- detecting whether the overlay rectangle (showing the screen content) origin is filled with valid values.

Manually tested.

* Source/WebCore/PAL/pal/mac/ScreenCaptureKitSoftLink.h:
* Source/WebCore/PAL/pal/mac/ScreenCaptureKitSoftLink.mm:
* Source/WebCore/platform/mediastream/mac/ScreenCaptureKitCaptureSource.h:
* Source/WebCore/platform/mediastream/mac/ScreenCaptureKitCaptureSource.mm:
(-[WebCoreScreenCaptureKitHelper outputVideoEffectDidStartForStream:]):
(-[WebCoreScreenCaptureKitHelper outputVideoEffectDidStopForStream:]):
(WebCore::ScreenCaptureKitCaptureSource::streamDidOutputVideoSampleBuffer):

Canonical link: <a href="https://commits.webkit.org/278390@main">https://commits.webkit.org/278390@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/8daea2d3986e2e6f8cd92b226637389c6fd24755

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/50294 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/29587 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/2591 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/53553 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/984 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/52597 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/35815 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/631 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/41031 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/52393 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/27244 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/43290 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/22135 "Passed tests") | 
| | [⏳ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/iOS-17-Simulator-WPT-WK2-Tests-EWS "Waiting to run tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/543 "Passed tests") | [✅ 🛠 wpe-skia](https://ews-build.webkit.org/#/builders/52/builds/8672 "Built successfully") | 
| | [⏳ 🧪 api-ios](https://ews-build.webkit.org/#/builders/API-Tests-iOS-Simulator-EWS "Waiting to run tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/607 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/55139 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/25390 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/540 "Passed tests") | [⏳ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/GTK-WK2-Tests-EWS "Waiting to run tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/26653 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/43473 "Passed tests") | [⏳ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/API-Tests-GTK-EWS "Waiting to run tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/11051 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/27515 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/26383 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

youennf commented on Source/WebCore/PAL/pal/mac/ScreenCaptureKitSoftLink.mm:
```diff
@@ -47,5 +47,10 @@
+SOFT_LINK_CONSTANT_FOR_SOURCE_WITH_EXPORT(PAL, ScreenCaptureKit, SCStreamFrameInfoBoundingRect, NSString *, PAL_EXPORT)
+
+#if HAVE(SC_CONTENT_SHARING_PICKER)
+SOFT_LINK_CONSTANT_FOR_SOURCE_WITH_EXPORT(PAL, ScreenCaptureKit, SCStreamFrameInfoPresenterOverlayContentRect, NSString *, PAL_EXPORT)
```
As suggested by Eric:
```suggestion
SOFT_LINK_CONSTANT_MAY_FAIL_FOR_SOURCE_WITH_EXPORT(PAL, ScreenCaptureKit, SCStreamFrameInfoPresenterOverlayContentRect, NSString *, PAL_EXPORT)
```

youennf commented on Source/WebCore/PAL/pal/mac/ScreenCaptureKitSoftLink.h:
```diff
@@ -52,4 +52,9 @@ SOFT_LINK_CONSTANT_FOR_HEADER(PAL, ScreenCaptureKit, SCStreamFrameInfoContentSca
 #define SCStreamFrameInfoContentRect PAL::get_ScreenCaptureKit_SCStreamFrameInfoContentRect()
 
+#if HAVE(SC_CONTENT_SHARING_PICKER)
+SOFT_LINK_CONSTANT_FOR_HEADER(PAL, ScreenCaptureKit, SCStreamFrameInfoPresenterOverlayContentRect, NSString *)
```
As suggested by Eric:
```suggestion
SOFT_LINK_CONSTANT_MAY_FAIL_FOR_HEADER(PAL, ScreenCaptureKit, SCStreamFrameInfoPresenterOverlayContentRect, NSString *)
```

eric-carlson commented on Source/WebCore/platform/mediastream/mac/ScreenCaptureKitCaptureSource.mm:
```diff
@@ -506,7 +534,15 @@ - (void)stream:(SCStream *)stream didOutputSampleBuffer:(CMSampleBufferRef)sampl
-
+#if HAVE(SC_CONTENT_SHARING_PICKER)
+        if (m_isVideoEffectEnabled) {
+            if (auto overlayRectDictionary = (CFDictionaryRef)attachment[SCStreamFrameInfoPresenterOverlayContentRect]) {
```
You need to check that the constant is avaiable, something like:

```
    auto canCheckForOverlayMode = PAL::canLoad_ScreenCaptureKit_SCStreamFrameInfoPresenterOverlayContentRect();
    [attachments enumerateObjectsUsingBlock:makeBlockPtr([&] (NSDictionary *attachment, NSUInteger, BOOL *stop) {

...

        if (canCheckForOverlayMode  && m_isVideoEffectEnabled) {
```

eric-carlson committed:
Do not trigger stream configuration updates in case user is capturing screen in large presenter overlay mode

Do not trigger stream configuration updates in case user is capturing screen in large presenter overlay mode
https://bugs.webkit.org/show_bug.cgi?id=273684
rdar://125925090

Reviewed by Eric Carlson.

We enter in a loop of reconfiguration when trying to update the stream configuration size when user selects large presenter overlay mode.
We are now skipping the reconfiguration step.
This reintroduces black stripes like there used to have before rdar://124131045.
A future patch will fix this by adding cropping within ScreenCaptureKitCaptureSource.

We detect large presenter overlay mode by:
- detecting whether video effect is enabled (which means presenter mode is on or off)
- detecting whether the overlay rectangle (showing the screen content) origin is filled with valid values.

Manually tested.

* Source/WebCore/PAL/pal/mac/ScreenCaptureKitSoftLink.h:
* Source/WebCore/PAL/pal/mac/ScreenCaptureKitSoftLink.mm:
* Source/WebCore/platform/mediastream/mac/ScreenCaptureKitCaptureSource.h:
* Source/WebCore/platform/mediastream/mac/ScreenCaptureKitCaptureSource.mm:
(-[WebCoreScreenCaptureKitHelper outputVideoEffectDidStartForStream:]):
(-[WebCoreScreenCaptureKitHelper outputVideoEffectDidStopForStream:]):
(WebCore::ScreenCaptureKitCaptureSource::streamDidOutputVideoSampleBuffer):

Canonical link: https://commits.webkit.org/278390@main
